### PR TITLE
add "kd metrics" subcommand 

### PR DESCRIPTION
**Pull request by cihangir**
~~deps on https://github.com/koding/koding/pull/11132~~

```
➜  klientctl git:(kd_metrics_3) ✗ go test . -v -bench . -benchmem -run=MetricsOverhead
BenchmarkMetricsOverheadTags-8   	    5000	    350790 ns/op	   42238 B/op	     511 allocs/op
BenchmarkMetricsOverheadSend-8   	    3000	    454371 ns/op	   78524 B/op	     149 allocs/op
BenchmarkMetricsOverheadAll-8    	    1000	   1099093 ns/op	  160998 B/op	     747 allocs/op
```

### CODE REVIEW

File: go/src/koding/kites/kloud/kloud/kloud.go [hunk 1]
```diff
@@ -413,16 +409,46 @@ func (k *Kloud) HandleFunc(pattern string, f kite.HandlerFunc) *kite.Method {
...
+		}
 	}
 
-	k.Kite.Close()
-	k.Stats.Close()
+	if err := k.Stats.Close(); err != nil {
+		merr = multierror.Append(merr, err)
+	}
+
+	defer k.closeOnce.Do(func() {
```
Comment (rjeczalik) [inline go/src/koding/kites/kloud/kloud/kloud.go]: defer seems not needed here

File: go/src/koding/kites/kloud/kloud/kloud.go [hunk 2]
```diff
@@ -413,16 +409,46 @@ func (k *Kloud) HandleFunc(pattern string, f kite.HandlerFunc) *kite.Method {
...
+		signal.Notify(signalCh)
+
+		s := <-signalCh
+		signal.Stop(signalCh)
+		switch s {
+		case syscall.SIGINT, syscall.SIGTERM, syscall.SIGKILL:
+			k.Kite.Log.Info("%s signal received, closing kloud", s)
+			k.Close()
+		}
+	}()
```
Comment (rjeczalik) [inline go/src/koding/kites/kloud/kloud/kloud.go]: the caller could do `go k.handleSignals()` so we could remove the go / closure from here

File: go/src/koding/klientctl/main.go [hunk 3]
```diff
@@ -145,12 +145,42 @@ func run(args []string) {
 
 	signal.Notify(sig, signals...)
 
+	m, err := metrics.New("kd")
+	if err != nil {
+		fmt.Fprintln(os.Stderr, "metrics wont be collected: ", err)
+	}
+	defer m.Close()
```
Comment (rjeczalik) [inline go/src/koding/klientctl/main.go]: if m != nil { defer m.Close() }

**Comment by rjeczalik** [comment]
Maybe stupid question to ask, but why we need metrics subcommand?

File: go/src/koding/kites/kloud/kloud/kloud.go [hunk 4]
```diff
@@ -413,16 +409,46 @@ func (k *Kloud) HandleFunc(pattern string, f kite.HandlerFunc) *kite.Method {
...
+		signalCh := make(chan os.Signal, 1)
+		signal.Notify(signalCh)
+
+		s := <-signalCh
+		signal.Stop(signalCh)
+		switch s {
+		case syscall.SIGINT, syscall.SIGTERM, syscall.SIGKILL:
+			k.Kite.Log.Info("%s signal received, closing kloud", s)
+			k.Close()
+		}
```
Comment (ppknap) [inline go/src/koding/kites/kloud/kloud/kloud.go]: What about:
```go
for {
	switch s := <-signalCh; s {
	case syscall.SIGINT, syscall.SIGTERM, syscall.SIGKILL:
		k.Kite.Log.Info("%s signal received, closing kloud", s)
		signal.Stop(signalCh)
		k.Close()
		return
	}
}
```
If we catch different signal than `syscall.SIGINT, syscall.SIGTERM, syscall.SIGKILL`, this go-routine will exit as no-op.

We can also replace `switch` with `select` and catch `k.closeChan` to not leak goroutines.

File: go/src/koding/klientctl/metrics_test.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+import (
+	"koding/kites/metrics"
+	"testing"
+	"time"
+
+	"github.com/codegangsta/cli"
+)
+
+func BenchmarkMetricsOverheadTags(b *testing.B) {
```
Comment (ppknap) [inline go/src/koding/klientctl/metrics_test.go]: 💯

File: go/src/koding/kites/kloud/kloud/kloud.go (hunk 4 above)
Comment (rjeczalik) [inline go/src/koding/kites/kloud/kloud/kloud.go]: Usually second signal forcefully closes a process, e.g. sending CTRL+C twice. With for-loop it won't never stop, unless -9 is sent.

**Comment by cihangir** [comment]
https://github.com/koding/koding/pull/11142#issuecomment-301006189

Atom will use that command to send their internal metrics to our system.

File: go/src/koding/kites/kloud/kloud/kloud.go (hunk 4 above)
Comment (ppknap) [inline go/src/koding/kites/kloud/kloud/kloud.go]: I'm talking about other signals like `SIGUSR1` etc. With current logic we will catch it, `switch-case` will not be executed end the function will exit.

Comment (rjeczalik) [inline go/src/koding/kites/kloud/kloud/kloud.go]: I'm talking about the code snippet you posted, not sure about SIGUSR1 etc.

Comment (ppknap) [inline go/src/koding/kites/kloud/kloud/kloud.go]: Or just better solution:
```go
func Notify(signalCh, syscall.SIGINT, syscall.SIGTERM, syscall.SIGKILL)
```
without `switch`

**Comment by rjeczalik** [comment]
> Atom will use that command to send their internal metrics to our system.

I missed that one, good to know.

File: go/src/koding/kites/kloud/kloud/kloud.go (hunk 4 above)
Comment (cihangir) [inline go/src/koding/kites/kloud/kloud/kloud.go]: https://github.com/koding/koding/pull/11142#discussion_r116258392

We will only get [SigTerm](https://github.com/koding/koding/blob/master/deployment/supervisord.coffee#L54) 
So we are on safe side

**Review by ppknap — APPROVED** [review]

**Comment by cihangir** [comment]
Build is passing now

**Review by rjeczalik — APPROVED** [review]

**Pull request merged**